src/data/ProgramsD2Repository.ts

```diff
@@ -87,20 +101,109 @@ export class ProgramsD2Repository implements ProgramsRepository {
         return d2ProgramRules.run(options);
     }
 
+    async getAll(options: MoveProgramAttributeOptions): Async<ProgramAttributes[]> {
+        const trackedEntities = await this.getFromTracker<D2TrackedEntity>("trackedEntities", {
+            orgUnitIds: undefined,
+            programIds: [options.programId],
+        });
+
+        return trackedEntities.map(tei => {
+            return {
+                id: tei.trackedEntity,
+                orgUnit: tei.orgUnit,
+                trackedEntityType: tei.trackedEntityType,
+                attributes: tei.attributes.map(attribute => {
+                    return {
+                        attribute: attribute.attribute,
+                        value: attribute.value,
+                        storedBy: attribute.storedBy,
+                    };
+                }),
+                programId: options.programId,
+            };
+        });
+    }
+
+    async saveAttributes(programs: ProgramAttributes[]): Async<Stats> {
+        if (programs.length === 0)
+            return { created: 0, ignored: 0, updated: 0, errorMessage: "", recordsSkipped: [] };
+        const teisToFetch = programs.map(program => program.id);
+        const programsIds = _(programs)
+            .map(program => program.programId)
+            .uniq()
+            .value();
+
+        const programsByKey = _(programs)
+            .keyBy(tei => tei.id)
+            .value();
+
+        const stats = await getInChunks<Stats>(teisToFetch, async teiIds => {
+            const trackedEntities = await this.getFromTracker<D2TrackedEntity>("trackedEntities", {
+                orgUnitIds: undefined,
+                programIds: programsIds,
+                trackedEntity: teiIds.join(";"),
+            });
+
+            const teisToSave = trackedEntities.map(tei => {
+                const currentProgram = programsByKey[tei.trackedEntity];
+                if (!currentProgram) throw Error(`Cannot find program: ${tei.trackedEntity}`);
+                const attributes = currentProgram?.attributes || tei.attributes;
```

Comment:
There are two `currentProgram?` but we have a guard for the undefined value (which is correct), so the code can be simplified.